compiler/optimizer/abstractinterpreter/AbsValue.hpp

```diff
@@ -0,0 +1,160 @@
+    * Any type promotion, such as Int16 => Int32 should be handled outside before calling merge() since this is not handled inside this method.
+    *
+    * @param other AbsValue*
+    * @param vp OMR::ValuePropagation* 
+    * @return AbsValue*
+    * 
+    * The return value is self or NULL.
+    * When succeeding to merge, it returns self.
+    * NULL denotes the merge fails (when merging with different dataTypes or merging with a Dummy AbsValue).
+    */
+   AbsValue* merge(AbsValue *other, OMR::ValuePropagation *vp);
+
+   /**
+    * @brief Check whether the AbsValue has no constraint
+    *
+    * @return bool
+    */
+   bool isTop() { return _constraint == NULL; }
+
+   /**
+    * @brief Set the constriant to unknown.
+    *
+    * @return void
+    */
+   void setToTop() { _constraint = NULL; }
+
+   /**
+    * @brief Check if the AbsValue is only used to take a slot as the second word of the 64-bit data types
+    *
+    * @return bool
+    */
+   bool isDummy() { return _isDummy; }
+
+   /**
+    * @brief Check if the AbsValue is a parameter.
+    *
+    * @return bool
+    */
+   bool isParameter() { return _paramPos >= 0; }
+
+   /**
+    * @brief Check if the AbsValue is 'this' (the implicit parameter.)
+    *
+    * @return bool
+    */   
+   bool isImplicitParameter() { return _paramPos == 0 && _isImplicitParameter; }
+
+   /**
+    * A set of methods for creating different kinds of AbsValue.
+    * OMR::ValuePropagation is required as a parameter since we are using TR::VPConstraint as the constraint.
+    * Though it has nothing to do with the Value Propagation optimization.
+    */
+   static AbsValue* create(TR::VPConstraint *constraint, TR::DataType dataType, TR::Region& region);
+
+   static AbsValue* create(AbsValue* other, TR::Region& region);
+   
+   static AbsValue* createClassObject(TR_OpaqueClassBlock* opaqueClass, bool mustBeNonNull, TR::Region& region, OMR::ValuePropagation* vp);
+
+   static AbsValue* createNullObject(TR::Region& region, OMR::ValuePropagation* vp);
+   static AbsValue* createArrayObject(TR_OpaqueClassBlock* arrayClass, bool mustBeNonNull, int32_t lengthLow, int32_t lengthHigh, int32_t elementSize, TR::Region& region, OMR::ValuePropagation* vp);
+   
+   static AbsValue* createStringConst(TR::SymbolReference* symRef, TR::Region& region, OMR::ValuePropagation* vp);
+
+   static AbsValue* createIntConst(int32_t value, TR::Region& region, OMR::ValuePropagation* vp);
+   static AbsValue* createLongConst(int64_t value, TR::Region& region, OMR::ValuePropagation* vp);
+   static AbsValue* createIntRange(int32_t low, int32_t high, TR::Region& region, OMR::ValuePropagation* vp);
+   static AbsValue* createLongRange(int64_t low, int64_t high, TR::Region& region, OMR::ValuePropagation* vp);
+
+   static AbsValue* createTopInt(TR::Region& region);
+   static AbsValue* createTopLong(TR::Region& region);
+
+   static AbsValue* createTopObject(TR::Region& region);
```

Comment:
Does that really need to be clarified? Can't it just follow whatever we use in VP and elsewhere? The APIs can be overridden if the default 'object' isn't right for your language.

Reply:
I didn't express clearly my thinking here.

In this case specifically (though I believe there are similar cases), it looks to me like an unconstrained object is really an unconstrained _reference_. If that's correct, I don't think we should be implying that a reference must be an object, since that is generally not true. If that's not correct, then the implication is that "Object" refers to something more specific that wasn't obvious to me.

Reply:
So I think the key point for @xiacijie - is this top for any pointer or top for Objects and do we need a top for a reference that isn't an object. That seems a fair and reasonable question - thanks for the clarification.

Reply:
It is 'object'. The current implementation of the abstract interpretation does not model the heap so reference/pointer is not really useful.

Reply:
@xiacijie If I understand correctly, abstract interpreters have to be language-specific and must therefore be implemented by the user. Given that some languages may need to model a heap (or even non-heap memory), why wouldn't references/pointers be useful?